internal/cms/signed.go

```diff
@@ -0,0 +1,316 @@
+		return nil, ErrCertificateNotFound
+	}
+
+	intermediates := x509.NewCertPool()
+	for _, cert := range d.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	verifiedSignerMap := map[string]*x509.Certificate{}
+	var lastErr error
+	for _, signerInfo := range d.SignerInfos {
+		signingCertificate := d.GetCertificate(signerInfo.SignerIdentifier)
+		if signingCertificate == nil {
+			lastErr = ErrCertificateNotFound
+			continue
+		}
+
+		cert, err := d.VerifySigner(ctx, &signerInfo, signingCertificate, opts)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+
+		thumbprint, err := hashutil.ComputeHash(crypto.SHA256, cert.Raw)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		verifiedSignerMap[hex.EncodeToString(thumbprint)] = cert
+	}
+	if len(verifiedSignerMap) == 0 {
+		return nil, lastErr
+	}
+
+	verifiedSigningCertificates := make([]*x509.Certificate, 0, len(verifiedSignerMap))
+	for _, cert := range verifiedSignerMap {
+		verifiedSigningCertificates = append(verifiedSigningCertificates, cert)
+	}
+	return verifiedSigningCertificates, nil
+}
+
+// VerifySigner verifies the signerInfo against the signingCertificate.
+//
+// SignedData's certificates field is optional, so this function can be used
+// to verify the signerInfo without the certificates field. Or the user doesn't
+// trust the certificates of SignedData and the signer identifier of
+// the SignerInfo (they are unsigned fields), this function can be used to
+// verify the signerInfo against the user provided signingCertificate.
+//
+// Note: the intermediate certificates and root certificates in the verify
+// options should be set by the user. The certificates in the SignedData
+// will be ignored.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) VerifySigner(ctx context.Context, signerInfo *SignerInfo, signingCertificate *x509.Certificate, opts x509.VerifyOptions) (*x509.Certificate, error) {
+	if signerInfo == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signer info is required"}
+	}
+
+	if signingCertificate == nil {
+		return nil, VerificationError{Message: "VerifySigner failed: signing certificate is required"}
+	}
+
+	if signerInfo.Version != 1 {
+		// Only IssuerAndSerialNumber is supported currently
+		return nil, VerificationError{Message: fmt.Sprintf("invalid signer info version: only version 1 is supported; got %d", signerInfo.Version)}
+	}
+
+	// user provided signing certificate must match the one in signer info
+	if !bytes.Equal(signingCertificate.RawIssuer, signerInfo.SignerIdentifier.Issuer.FullBytes) || signingCertificate.SerialNumber.Cmp(signerInfo.SignerIdentifier.SerialNumber) != 0 {
+		return nil, SyntaxError{Message: "signing certificate does not match signer info"}
+	}
```

Comment:
On a second thought, I think we can remove this enforcement. RFC 5652 doesn't regulate the verification process of a cms signed-data though. Quote from RFC 5652, "...The recipient MAY obtain the correct public key for the signer by any means...". Since the caller of this function might not trust the sid field (it's unsigned), it should not be used as a check in this function.
The `Verify` function would still be correct, because it gets the cert using sid before calling `VerifySigner`. In other words, as one of the callers, the `Verify` function chooses to trust sid.
```suggestion
```